wallet/src/key_chain.rs

```diff
@@ -0,0 +1,186 @@
+    /// This is for the internal usage of the wallet when creating change output for a transaction
+    Change = 1,
+}
+
+#[allow(dead_code)] // TODO remove
+/// This key chain contains a pool of pre-generated keys and addresses for the usage in a wallet
+pub struct KeyChain<B: Backend> {
+    /// The specific chain this KeyChain is based on, this will affect the address format
+    chain_config: Arc<ChainConfig>,
+
+    /// The master key of this key chain from where all the keys are derived from
+    // TODO implement encryption
+    master_key: ExtendedPrivateKey,
+
+    /// The database connection
+    // TODO Consider if this should be an Option
+    db: Arc<Store<B>>,
+
+    /// The number of addresses to pre-generate from the master key
+    lookahead_size: u16,
+}
+#[allow(dead_code)] // TODO remove
+impl<B: Backend> KeyChain<B> {
+    pub fn new_from_mnemonic(
+        chain_config: Arc<ChainConfig>,
+        mnemonic_str: &str,
+        passphrase: Option<&str>,
+        db: Arc<Store<B>>,
+    ) -> KeyChainResult<Self> {
+        // TODO use Mnemonic::parse when bip39 is configured with "std"
+        let mnemonic =
+            bip39::Mnemonic::parse_normalized(mnemonic_str).map_err(KeyChainError::Bip39)?;
+        // TODO use Mnemonic::to_seed when bip39 is configured with "std"
+        let mut seed = mnemonic.to_seed_normalized(passphrase.unwrap_or(""));
+        let master_key = ExtendedPrivateKey::new_master(&seed, KEY_KIND)?;
+        // TODO(SECURITY) erase mnemonic
+        seed.zeroize();
+
+        Ok(KeyChain {
+            chain_config,
+            master_key,
+            db,
+            lookahead_size: LOOKAHEAD_SIZE,
+        })
+    }
+
+    pub fn load_key_chain(db: Arc<Store<B>>) -> KeyChainResult<Self> {
+        // TODO remove this
+        let _ = db.get_storage_version().expect("This should work?");
+        // TODO implement loading from database
+        Ok(KeyChain {
+            chain_config: Arc::new(create_regtest()),
+            master_key: ExtendedPrivateKey::new_from_entropy(KEY_KIND).0,
+            db,
+            lookahead_size: LOOKAHEAD_SIZE,
+        })
+    }
+
+    /// Get a new address that hasn't been used before
+    pub fn get_new_address(&self, purpose: KeyPurpose) -> KeyChainResult<Address> {
+        let key = self.get_new_key(purpose)?;
+
+        let address = Address::from_public_key(
+            &self.chain_config,
+            &ExtendedPublicKey::from_private_key(&key).public_key(),
+        )?;
+
+        // TODO save address
+
+        Ok(address)
+    }
+
+    /// Get a new derived key that hasn't been used before
+    pub fn get_new_key(&self, purpose: KeyPurpose) -> KeyChainResult<ExtendedPrivateKey> {
+        // TODO implement with correct paths
+        let hd_path = match purpose {
+            KeyPurpose::ReceiveFunds => DerivationPath::from_str("m/0'/0'/0'/0'")?,
+            KeyPurpose::Change => DerivationPath::from_str("m/0'/0'/1'/0'")?,
+        };
```

Comment:
Could the derivation path be passed in as a slice of indices rather than string? Avoids potential parsing issues.

Reply:
Those will be created from ChildNumber directly.